1

```diff
@@ -0,0 +1,87 @@
+import assert from 'assert';
+import axios from 'axios';
+import qs from 'qs';
+
+export const ENDPOINTS = {
+  REPORT_DATA: 'lib/data.php',
+  REPORT_CONFIG: 'lib/settings.php',
+  DATA_DICTIONARY: 'lib/data-dictionary.php'
+};
+
+/**
+ * Constructs a service that fetches data for the current project from REDCap.
+ *
+ * @param {Object} assetUrls - key/value pairs where the key is a relative file path
+ *   like `lib/data.php`, and the value is the corresponding external module URL
+ *   like `http://localhost/redcap/api/?type=module&prefix=consort-report&page=lib%2Fdata&pid=42`
+ * @return {Object} an object encapsulating REDCap HTTP requests
+ */
+export default function createDataService(assetUrls) {
+  assert(assetUrls, 'Asset URL object is required');
+  Object.keys(ENDPOINTS).forEach(key => {
+    const endpoint = ENDPOINTS[key];
+    assert(assetUrls[endpoint], `A URL for ${endpoint} is required`)
+  });
+
+  return {
+    reportDataUrl: assetUrls[ENDPOINTS.REPORT_DATA],
+    reportConfigUrl: assetUrls[ENDPOINTS.REPORT_CONFIG],
+    dataDictionaryUrl: assetUrls[ENDPOINTS.DATA_DICTIONARY],
+
+    /**
+     * Extracts data returned by the request.
+     * @param {Promise->Object} response - an axios response object.
+     * @return {Promise->Object} the response's data
+     * @private
+     */
+    _extractData(response) {
+      return response.data;
+    },
+
+    /**
+     * Get a report summary that includes the total number of records for a report.
+     *
+     * @return {Promise} A promise that resolves to an object with the following keys:
+     *   - reportId: The report ID.
+     *   - totalRecords: The total number of records for a report.
+     */
+    fetchReportSummary() {
+      return axios.post(this.reportDataUrl).then(this._extractData);
+    },
+
+    /**
+     * Post report summary for saving to report config.
+     * @param {Object} reportSummary Data for a report summary.
+     * @return {Promise->Object} the response's report summary data
+     */
+    saveReportSummary(reportSummary) {
+      const data = { reportSummary: JSON.stringify({ reportSummary }) };
+      const options = { timeout: 5000 };
```

Comment:
This is a nice touch. We haven't been very consistent about adding timeouts in the past, and it would be good to make it a habit.

What happens if the timeout expires? Does the promise reject?

Reply:
I'm assuming it will, but that should probably be checked/tested.

Reply:
Confirmed the timeout is handled. User receives a message like this:

![screen shot 2018-11-06 at 08 31 12](https://user-images.githubusercontent.com/273863/48078710-d2a65a80-e19e-11e8-950e-095a59156ef5.png)